### PR TITLE
Add yaml metadata parsing

### LINKED ISSUE #66: Feature suggestion: Support GFM YAML metadata block
Some Markdown implementations, e.g., GFM and Pandoc, support YAML metadata blocks. For GFM (according to testing, since it seems undocumented) the YAML block must start at the beginning at the document (even blank lines before it are not allowed) and be wrapped in `---` (*exactly* three dashes) at the opening and closing, so an example document would be

```markdown
---
title: README
author: Zhiming Wang
date: January 8, 2016
---

This is an example GFM document with a YAML metadata block.
```

and this will be rendered like

screen shot 2016-01-08 at 12 21 10 pm

Note that the blank line after the closing `---` line is optional.

(Pandoc, on the other hand, has very lax rules for YAML metadata blocks, which allows them to be placed anywhere in the document, and allows closing with `...` instead of `---`. But Pandoc Markdown doesn't need to be supported.)

Currently, the line immediately above the closing `---` is interpreted as a setext header, which looks rather weird (especially if you have a real header):

screen shot 2016-01-08 at 12 22 42 pm

Would you please consider supporting the metadata block? I suppose the block itself could be highlighted just like a fenced code block, and I suppose it shouldn't be too hard to adapt the parsing code of fenced code blocks to this particular block. Thanks.

### PR DESCRIPTION
**Pull request by cosmicexplorer**
Also add testing and allow multiple pandoc metadata blocks. Fix #66.

### CODE REVIEW

File: markdown-mode.el [hunk 1]
```diff
@@ -2599,30 +2633,60 @@ analysis."
         (set-match-data (list beg (point)))
         t))))
 
-(defun markdown-match-generic-metadata (regexp last)
+(defun markdown-get-match-boundaries (start-header end-header last &optional pos)
+  (save-excursion
+    (goto-char (if pos pos (point-min)))
```
Comment (syohex) [inline markdown-mode.el]: How about `(or pos (point-min))` like other parts ?

File: markdown-mode.el [hunk 2]
```diff
@@ -2599,30 +2633,60 @@ analysis."
...
+    (cl-loop
+     with cur-result = nil
+     while (and (< (point) last)
+                (re-search-forward (or start-header "\\`") last t)
+                (progn
+                  (setq cur-result (match-data))
+                  (re-search-forward
+                   (or end-header "\n\n\\|\n\\'\\|\\'") nil t)))
```
Comment (syohex) [inline markdown-mode.el]: Following expressions are loop invariants. They can be moved outside the loop.
-  `(or start-header "\\`")`
- `(or end-header "\n\n\\|\n\\'\\|\\'")`

File: markdown-mode.el [hunk 3]
```diff
@@ -2599,30 +2633,60 @@ analysis."
...
-                          (match-beginning 0)
-                        (point-max)))))
-    (cond ((>= (point) header-end)
-           ;; Don't match anything outside of the header.
+  ;; if start-header is nil, we assume metadata can only occur at the very top
+  ;; of a file ("\\`"). if end-header is nil, we assume it is "\n\n"
+  (let* ((header-bounds
+          (markdown-get-match-boundaries start-header end-header last))
+         (enclosing-header
+          (cl-first             ; just take first if multiple (maybe impossible)
```
Comment (syohex) [inline markdown-mode.el]: Can `cl-first` + `cl-remove-if-not` be replaced with `cl-find-if` ?

File: tests/markdown-test.el [hunk 4]
```diff
@@ -2270,7 +2271,8 @@ Base Header Level: 2  "
 
 (ert-deftest test-markdown-font-lock/mmd-metadata-after-header ()
   "Ensure that similar lines are not matched after the header."
-  (markdown-test-string "Title: peg-multimarkdown User's Guide  
+  (markdown-test-string
+   "Title: peg-multimarkdown User's Guide
```
Comment (syohex) [inline tests/markdown-test.el]: Please do not change unrelated part.

File: tests/markdown-test.el [hunk 5]
```diff
@@ -2282,7 +2284,8 @@ Base Header Level: 2  "
 
 (ert-deftest test-markdown-font-lock/pandoc-metadata ()
   "Basic Pandoc metadata tests."
-  (markdown-test-string "% title
+  (markdown-test-string
```
Comment (syohex) [inline tests/markdown-test.el]: Ditto

File: tests/markdown-test.el [hunk 6]
```diff
@@ -2252,7 +2252,8 @@ for (var i = 0; i < 10; i++) {
 
 (ert-deftest test-markdown-font-lock/mmd-metadata ()
   "Basic MultMarkdown metadata tests."
-  (markdown-test-string "Title: peg-multimarkdown User's Guide  
+  (markdown-test-string
+   "Title: peg-multimarkdown User's Guide
```
Comment (syohex) [inline tests/markdown-test.el]: Please do not change unrelated part.

**Comment by syohex** [comment]
Thanks for great work. I put some comments, please check them.

File: markdown-mode.el (hunk 1 above)
Comment (cosmicexplorer) [inline markdown-mode.el]: Fixed, not sure what I was thinking.

File: markdown-mode.el (hunk 2 above)
Comment (cosmicexplorer) [inline markdown-mode.el]: Also fixed.

File: markdown-mode.el (hunk 3 above)
Comment (cosmicexplorer) [inline markdown-mode.el]: Yes, that's what I was trying to do.

File: tests/markdown-test.el (hunk 6 above)
Comment (cosmicexplorer) [inline tests/markdown-test.el]: Removed.

**Comment by syohex** [comment]
Looks good. Could you squash commits ?

**Comment by cosmicexplorer** [comment]
Done.

**Comment by syohex** [comment]
And please remove following lines from commit messages, they makes no sense.

> - Add defvar for markdown-search-until-condition
> - Update with comments from @syohex

**Comment by cosmicexplorer** [comment]
Fixed.

**Comment by syohex** [comment]
Thanks :+1:

**Comment by jrblevin** [comment]
Thanks for this patch. I only noticed one potential issue: YAML blocks inside fenced code blocks should be ignored:

    ```
    ---
    title: foo
    author: bar
    ---
    ```

You can use `markdown-code-block-at-pos` or `markdown-code-block-at-point` in the propertize function to skip code blocks.

**Comment by jrblevin** [comment]
Regarding my comment above about YAML in code blocks, I appended the test case and marked it with `  :expected-result :failed` for now. See commit 90f3f59. I took a shot at fixing it, but it's more complicated than I thought. I think we also need to check for code blocks in `markdown-get-match-boundaries`, because of the separate matching functions.

**Comment by cosmicexplorer** [comment]
I'll take a look at these tomorrow. I think I can speed up the metadata block matching by propertizing metadata block start and end boundaries (the `---\n` lines) and searching for characters with that property. We can then search for code block boundaries as well, using the same trick.

With respect to the if statements, I wanted to allow for local binding of the pandoc style; if you want to remove the if statements, I can factor that out into a separate function.

**Comment by vyp** [comment]
Thank you @cosmicexplorer for this! :+1:

**Comment by cosmicexplorer** [comment]
Thinking about it a bit more, metadata would be invalid inside any other "block"-level construct, such as gfm code blocks or tilde blocks; I think that's it? All the other block constructs have some required characters at the beginning of each line. I think what would probably work would be to propertize all three "border" types (gfm blocks, pandoc tilde blocks, yaml metadata blocks) at once in a single pass, then figure out the appropriate context for any text enclosed within them by finding which if any characters with the syntax property for start/end blocks of those three block types are closest to the characters to propertize.

So basically, if our text is like:
```
... (some stuff without ~~~, ```, or ---)
`` ` (propertized as gfm-codeblock-begin by first block-construct pass)
--- (not propertized because already within block construct, but with markdown-inline-code-face)
key: value (look for enclosing block propertized text; found gfm-codeblock, so no propertizing/highlighting)
--- (not propertized)
`` ` (gfm-codeblock-end)
```
